app/core/models/abortables.ts

```diff
@@ -0,0 +1,54 @@
+import {nanoid} from "@reduxjs/toolkit"
+import {isString, remove} from "lodash"
+
+export type Abortable<Meta extends any = any> = {
+  id: string
+  tab?: string
+  tag?: string
+  meta?: Meta
+  abort: () => void | Promise<void>
+}
+
+type NewAbortable = Omit<Abortable, "id">
+
+export class Abortables {
+  registry: Abortable[] = []
+
+  abort(predicate?: Partial<Abortable>) {
+    this.filter(predicate).forEach((a) => a.abort())
+  }
+
+  add(a: Abortable | NewAbortable) {
+    const id = "id" in a ? a.id : nanoid()
+    this.registry.push({id, ...a})
+    return id
+  }
+
+  all() {
+    return this.registry
+  }
+
+  filter(predicate?: Partial<Abortable>) {
+    if (!predicate) return this.registry
```

Comment:
Safer to return a ref to a copy of the registry, instead of a ref to the registry itself when there is no predicate so it can't be tampered with externally. Then in the case where there is a predicate that's what filter is already doing.

Reply:
Agreed. I'll update this.